proposals/0000-ghc-cpp.rst

```diff
@@ -0,0 +1,231 @@
+Ghc CPP
+=======
+
+.. author:: Alan Zimmerman
+.. date-accepted:: 
+.. ticket-url::
+.. implemented:: 
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/616>`_.
+.. sectnum::
+.. contents::
+
+Haskell code uses the CPP pragma extensively. I'm proposing to
+introduce a new GhcCPP pragma to perform a subset of CPP processing as
+an alternative, as part of the normal GHC compilation process.
+
+This subset will be just enough to control conditional compilation,
+and will only be active in lines introduced with a ``#`` and
+preprocessor keyword.
+
+It will be tooling friendly, in that parts of the code disabled by the
+processor will be saved as comments, if comments are being processed.
+
+
+Motivation
+----------
+
+Haskell code uses the CPP pragma to manage version changes, when the
+compiler or a library changes.
+
+CPP gets the job done, but it is not haskell-friendly, and so has a
+side-effect of mangling otherwise normal haskell code, which is not
+intended to be processed for the version-selection process.
+
+It is also not tooling-friendly. It runs as a separate pass over the
+source code, emitting updated source for compilation which can
+actually change pretty much anything, if a person decides to use the
+full power of CPP to redefine existing Haskell, or introduce macros to
+generate boilerplate code.
+
+The primary motivation for this proposal is to make it possible to
+safely perform source-code transformations on code containing
+preprocessor directives.
+
+Currently in ``ghc-exactprint`` when a file uses CPP, the original
+file and preprocessor output are diffed at the token level and the
+"missing" parts inserted as comments. This works in some cases, but
+for many it fails, for reasons such as
+
+- A comparison to a numeric constant:
+
+.. code:: haskell
+
+   #if __GLASGOW_HASKELL__ < 707
+
+Shows up as a stray ``906`` if compiled with GHC 9.6.x
```

Comment:
```suggestion
Shows up as a stray ``906`` if compiled with GHC 9.6.x
```
The whole ``#if __GLASGOW_HASKELL__ < 707`` shows up as ``906`` or just the ``__GLASGOW_HASKELL__``?

Reply:
Using the `ghc-exactprint` diffing process, and printing,  this ends up something like

```
#if __GLASGOW_HASKELL__906 < 707
```

The point being it makes source-to-source tooling impossible